### PR TITLE
[1.19] Add container out of memory metrics

### PR DESCRIPTION
**Pull request by dadan2019**
Backport #4690 to 1.19
---------------------------------------------------------------
This patch adds a new metric
`container_runtime_crio_containers_oom_total` and
`container_runtime_crio_containers_oom`, which collects out of memory
(oom) metrics global and by container name. This also includes
sandboxes, since we reuse the CRI-O internal name for every container.

Signed-off-by: Sascha Grunert <sgrunert@redhat.com>

#### What this PR does / why we need it:
context: https://kubernetes.slack.com/archives/CAZH62UR1/p1638357795128600

#### Does this PR introduce a user-facing change?

```release-note
Added `container_runtime_crio_containers_oom_total` and  `container_runtime_crio_containers_oom` metrics,
which collects out of memory (oom) containers.
```

### CODE REVIEW

File: internal/oci/runtime_oci.go [hunk 1]
```diff
@@ -892,6 +893,17 @@ func (r *runtimeOCI) UpdateContainerStatus(c *Container) error {
...
 	if _, err = os.Stat(oomFilePath); err == nil {
 		c.state.OOMKilled = true
+
+		// Collect total metric
+		metrics.CRIOContainersOOMTotal.Inc()
+
+		// Collect metric by container name
+		counter, err := metrics.CRIOContainersOOM.GetMetricWithLabelValues(c.Name())
+		if err != nil {
+			log.Warnf(ctx, "Unable to write OOM metric by container: %v", err)
```
Comment (haircommander) [inline internal/oci/runtime_oci.go:903]: we haven't wired ctx throug hyet, so this should be `logrus.Warnf("Unable...")`

File: internal/oci/runtime_vm.go [hunk 2]
```diff
@@ -664,6 +665,17 @@ func (r *runtimeVM) updateContainerStatus(c *Container) error {
...
 		if _, err = os.Stat(oomFilePath); err == nil {
 			c.state.OOMKilled = true
+
+			// Collect total metric
+			metrics.CRIOContainersOOMTotal.Inc()
+
+			// Collect metric by container name
+			counter, err := metrics.CRIOContainersOOM.GetMetricWithLabelValues(c.Name())
+			if err != nil {
+				log.Warnf(ctx, "Unable to write OOM metric by container: %v", err)
```
Comment (haircommander) [inline internal/oci/runtime_vm.go:675]: same here

File: internal/oci/runtime_oci.go (hunk 1 above)
Comment (dadan2019) [reply internal/oci/runtime_oci.go]: Fixed

File: internal/oci/runtime_vm.go (hunk 2 above)
Comment (dadan2019) [reply internal/oci/runtime_vm.go]: Fixed

**Comment by haircommander** [comment]
I believe cherry-picking https://github.com/cri-o/cri-o/pull/5668/commits/e50405e5a39fcd3007791a9c02c0a013781584b4 will fix that, would you mind doing so here?

**Comment by haircommander** [comment]
sorry @dadan2019 one more nit: can you sign the commit ed1a738b6c7404b87bab34793546eb42e076aa21 (the second one fixing the ctx issue) or squash it into the previous commit? 

adding the line
Signed-off-by: $name <$email> 

to the commit should do it

**Comment by dadan2019** [comment]
@haircommander I started this PR on Monday and thank you for your response every day. I wonder if it would be possible to get it merged and have the release packages ready by the end of this week or next Monday. I have a work item which depends on this and I used up my time budget. Building a private package is not what I want to do unless I have no choices.  Appreciate your help.

**Comment by haircommander** [comment]
thank you for your patience @dadan2019 
/lgtm
/approve
/override ci/openshift-jenkins/integration_rhel

**Comment by dadan2019** [comment]
> 
@haircommander Thanks a lot for doing '/cherry-pick release-1.20' so I don't need to do it again in 1.20. Do know when the release package (1.19.5?) will be ready?

**Comment by haircommander** [comment]
the cherry-pick didn't actually work, but I can take care of it if you'd like :) 

I plan on cutting 1.19.5 early next week, there are a couple more PRs to get in

**Comment by dadan2019** [comment]
> the cherry-pick didn't actually work, but I can take care of it if you'd like :)
> 
> I plan on cutting 1.19.5 early next week, there are a couple more PRs to get in

Please go ahead to do it for 1.20. It would be much more efficient if you do it :). 
I am looking forward to downloading the 1.19.5 release package early next week!

**Comment by haircommander** [comment]
backported here https://github.com/cri-o/cri-o/pull/5706

**Pull request merged**